Commonize toolchain rpm extract flows

dmcilvaney opened the pull request:
<!--
COMMENT BLOCKS WILL NOT BE INCLUDED IN THE PR.
Feel free to delete sections of the template which do not apply to your PR, or add additional details
-->

###### Merge Checklist  <!-- REQUIRED -->
<!-- You can set them now ([x]) or set them later using the Github UI -->
**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of the changes. -->
The full end-to-end toolchain builds are triggering the fixup extract step for every single toolchain rpm (`$(toolchain_rpms):` target). The flow which sets `REBUILD_TOOLCHAIN=n TOOLCHAIN_ARCHIVE=toolchain.tar.gz` however uses a staging mechanism so that we only need to extract the files once (the current flow invokes tar for every rpm which is very slow)

These changes make both paths run through a common set of build targets, with a number of improvements. The archive file should now only be extracted once into a temporary folder. All subsequent operations just copy the toolchain rpms from that temporary staging directory into `out/RPMS` or the rpm cache.

Any changes to the toolchain archive or manifests etc should trigger a wipe and re-extraction into that folder.

###### Change Log  <!-- REQUIRED -->
<!-- Detail the changes made here. -->
<!-- Please list any packages which will be affected by this change, if applicable. -->
<!-- Please list any CVES fixed by this change, if applicable. -->
- Moved extract directory from ./build/toolchain_temp to ./build/toolchain/extract_dir
- Converted both toolchain population flows to use the same set of targets by way of setting the variable `selected_toolchain_archvie`:
  - `$(STATUS_FLAGS_DIR)/toolchain_local_temp.flag:` Which populates the extract directory with the contents of `selected_toolchain_archive`
  - `$(toolchain_rpms):` Which goes through and makes sure each rpm is present in its desired location (./out/RPMS or the cache directory)
    - Each .rpm is copied from the extract directory if it is out of date or missing (much quicker to copy and re-extract from the archive)
- Toolchain validate is now a blocking error (if the toolchain manifest doesn't match the archive exactly)
- Tooling now watches the REBUILD_TOOLCHAIN=? and TOOLCHAIN_ARCHIVE=? configs for changes and will re-extract the toolchain RPMs if they change
- Package generation now explicitly requires ALL toolchain rpms if `REBUILD_TOOLCHAIN=y`, so we can guarantee that we have all the rpms needed to resolve cycles.
- `build-packages` target also explicitly requires all toolchain rpms if `REBUILD_TOOLCHAIN=y`(seems more consistent, since we already guarantee the chroot rpms are there)
- Print a warning message suggesting the user run `copy-toolchain-rpms` in the event of a cycle (This is a very common gotcha if you skip a step building the core repo)

###### Does this affect the toolchain?  <!-- REQUIRED -->
<!-- Any packages which are included in the toolchain should be carefully considered. Make sure the toolchain builds with these changes if so. -->
<!-- Update: manifests/package/toolchain_*.txt, pkggen_core_*.txt, update_manifests.sh -->
<!-- To validate: make clean; make workplan REBUILD_TOOLCHAIN=y DISABLE_UPSTREAM_REPOS=y CONFIG_FILE="" ... -->
**YES**

dmcilvaney commented on toolkit/scripts/toolchain.mk:
```diff
@@ -144,86 +144,72 @@ $(raw_toolchain): $(toolchain_files)
 # Output:
 # out/toolchain/built_rpms
 # out/toolchain/toolchain_built_rpms.tar.gz
-$(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.flag
```
Note: Revert this

MateuszMalisz commented:
How much will this increase the disk usage during a build?

dmcilvaney commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
 		echo "$${diff}"; \
-		$(call print_warning, $@ failed) ; \
-	fi
+		$(call print_error, $@ failed) ; \
```
This will detect mismatches in the toolchain manifests (which can have nasty knock-on effects in other build stages). I set this to a warning when I first added it since we were working to stabilize 2.0 at the time. Now that things have calmed down a bit I think we should turn this on.

dmcilvaney commented:
Added a dependency on the toolchain state for the chroot tar.gz as well, so it gets regenerated if we swap our toolchain around

dmcilvaney commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
-$(STATUS_FLAGS_DIR)/toolchain_local_temp.flag: $(TOOLCHAIN_ARCHIVE) $(shell find $(toolchain_local_temp)/* 2>/dev/null)
-	mkdir -p $(BUILD_DIR)/toolchain_temp/ && \
-	tar -I $(ARCHIVE_TOOL) -xvf $(TOOLCHAIN_ARCHIVE) -C $(BUILD_DIR)/toolchain_temp/ --strip-components 1 && \
+# The $(depend_TOOLCHAIN_ARCHIVE) and $(depend_REBUILD_TOOLCHAIN) argument trackers are used to force a regeneration of the toolchain
```
It may not involve actually building anything, this just goes through and makes sure all the .rpm files we expect are available where they are meant to be. I'll add a better comment to that effect.

dmcilvaney commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
 	touch $@
+	@echo Done validating toolchain
 
+# Extract a set of RPMS from an archive instead of building them from scratch.
```
Makes sense to me

dmcilvaney commented:
> How much will this increase the disk usage during a build?

I've seen ~1.5 to 1.6 GB

dmcilvaney commented:
Running pipeline build 191856 to validate

christopherco commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
+# We know how to build this archive from scratch
+selected_toolchain_archive = $(final_toolchain)
+else
+# This may be an empty string, that is fine. If its empty we will try to use online packages
```
Perhaps flesh out this comment a bit more to describe the cases.

christopherco commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
-	tar -I $(ARCHIVE_TOOL) -xvf $(TOOLCHAIN_ARCHIVE) -C $(BUILD_DIR)/toolchain_temp/ --strip-components 1 && \
+
+# If $(depend_TOOLCHAIN_ARCHIVE) and $(depend_REBUILD_TOOLCHAIN) argument trackers change it is important to check
+#	that all of the toolchain .rpms are correct. The different toolchain sources may have identail files but with
```
```suggestion
#	that all of the toolchain .rpms are correct. The different toolchain sources may have identical files but with
```

christopherco commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
+	mkdir -p $(toolchain_local_temp) && \
+	rm -f $(toolchain_local_temp)/* && \
+	tar -I $(ARCHIVE_TOOL) -xf $(selected_toolchain_archive) -C $(toolchain_local_temp) --strip-components 1 && \
 	touch $(BUILD_DIR)/toolchain_temp/* && \
```
```suggestion
	touch $(toolchain_local_temp)/* && \
```

christopherco commented on toolkit/scripts/tools.mk:
```diff
@@ -146,7 +146,7 @@ worker_chroot_deps := \
 
 ifeq ($(REFRESH_WORKER_CHROOT),y)
-$(chroot_worker): $(worker_chroot_deps)
+$(chroot_worker): $(worker_chroot_deps) $(depend_REBUILD_TOOLCHAIN) $(depend_TOOLCHAIN_ARCHIVE)
```
Possibly add a comment for `$(worker_chroot_deps)` for how it inspects the toolchain cache

anphel31 commented on toolkit/scripts/toolchain.mk:
```diff
@@ -171,59 +171,72 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
+endif
+
+# We will have three options at this point:
+# 1) REBUILD_TOOLCHAIN: yes										-> Rebuild a toolchain from scratch and pace it into $(final_toolchain)
```
can you align the spacing here? also typo: pace -> place

dmcilvaney replied:
Oh, interesting, GitHub is assuming a different tab size for this file. Makefiles generally use tabs everywhere, but I'll convert the spacing in those lines to spaces since they are comments.

anphel31 reviewed (commented):
I'm seeing an issue with "make copy-toolchain-rpms". Might need to update this as well
```
sudo make copy-toolchain-rpms
...
cp: cannot stat '/CBL-Mariner/build/rpm_cache/cache/x86_64/zstd-doc-1.5.0-1.cm2.x86_64.rpm': No such file or directory
toolkit/scripts/toolchain.mk:67: recipe for target 'copy-toolchain-rpms' failed
make: *** [copy-toolchain-rpms] Error 1
```

dmcilvaney commented:
> I'm seeing an issue with "make copy-toolchain-rpms". Might need to update this as well
> 
> ```
> sudo make copy-toolchain-rpms
> ...
> cp: cannot stat '/CBL-Mariner/build/rpm_cache/cache/x86_64/zstd-doc-1.5.0-1.cm2.x86_64.rpm': No such file or directory
> toolkit/scripts/toolchain.mk:67: recipe for target 'copy-toolchain-rpms' failed
> make: *** [copy-toolchain-rpms] Error 1
> ```

Was this from local testing? I wasn't able to repro it on my end. I think this is what it looks like if you run this command after doing a local build rather than a copy from archive/web build.

dmcilvaney commented:
Added dependencies to catch the edge case where one or more toolchain rpms are missing, but are NOT part of the chroot. The tools would not regenerate those missing rpms, since the only target we actually invoke are the chroot targets.

We need those rpms present to resolve circular dependencies in the graph.

I feel its outside the scope of the Makefile to be responsible for handling `copy-toolchain-rpms` etc automatically when you are not rebuilding the toolchain from source but are working in the core repo. I've added a warning message for when we find circular dependencies that suggests the most common solution to the issue.
